File: ACL Beginner Contest/C - Connect Cities.py
import sys
sys.setrecursionlimit(10000000)

N, M = map(int, sys.stdin.readline().split())

graph = [[] for _ in range(N+10)]
visited = [False for _ in range(N+10)]

for _ in range(M):
    a, b = map(int, sys.stdin.readline().split())
    graph[a-1].append(b-1)
    graph[b-1].append(a-1)

def dfs(idx : int):
    visited[idx] = True
    for i in graph[idx]:
        if not visited[i]:
            dfs(i)

cnt = 0

for i in range(N):
    if not visited[i]:
        dfs(i)
        cnt += 1

print(cnt - 1)

File: ACL Beginner Contest/C - Connect Cities.cpp
#include <bits/stdc++.h>
using namespace std;

int N, M, cnt;
vector<vector<int>> graph;
vector<int> visited;

void dfs(int idx)
{
    visited[idx] = 1;
    for (auto i : graph[idx])
        if (!visited[i])
            dfs(i);
}

int main()
{
    scanf("%d %d", &N, &M);
    graph.resize(N + 1);
    visited.resize(N + 1, 0);
    for (int i = 0, a, b; i < M; i++)
    {
        scanf("%d %d", &a, &b);
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    for (int i = 1; i <= N; i++)
    {
        if (!visited[i])
        {
            dfs(i);
            cnt++;
        }
    }
    printf("%d", cnt - 1);
    return 0;
}